src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1419,6 +1595,266 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+                    }
+                    catch (std::exception const& e)
+                    {
+                        JLOG(j_.error())
+                            << "shard " << shardIndex
+                            << " failed to move to historical storage";
+
+                        return;
+                    }
+
+                    // Create a shard instance at the new location
+                    shard = std::make_unique<Shard>(
+                        app_, *this, shardIndex, j_, dst);
+
+                    // Open the new shard
+                    if (!shard->open(scheduler_, *ctx_))
+                    {
+                        JLOG(j_.error())
+                            << "shard " << shardIndex
+                            << " failed to open in historical storage";
+
+                        shard->removeOnDestroy();
+                        shard.reset();
+                    }
+                }
+                else
+                {
+                    JLOG(j_.warn())
+                        << "can't find shard to move to historical path";
+                }
+            }
+            else
+            {
+                JLOG(j_.warn())
+                    << "can't find shard to move to historical path";
+            }
+        };
+        // Lambdas
+        //////////////////////////////////////////////////////////////////
+
+        // See if either of the recent shards
+        // needs to be updated
+        bool const curNotSynched =
+            latestShardIndex_ && *latestShardIndex_ != latestShardIndex;
+        bool const prevNotSynched = secondLatestShardIndex_ &&
+            *secondLatestShardIndex_ != latestShardIndex - 1;
+
+        // A new shard has been published. Move outdated shards
+        // to historical storage as needed
+        if (curNotSynched || prevNotSynched)
+        {
+            if (prev)
+            {
+                // Move the formerly second latest shard to
+                // historical storage
+                if (keepShard(*prev) && separateHistoricalPath)
+                {
+                    moveShard(*prev);
+                }
+
+                prev = boost::none;
+            }
+
+            if (cur)
+            {
+                // The formerly latest shard is now
+                // the second latest
+                if (cur == latestShardIndex - 1)
+                {
+                    prev = cur;
+                }
+
+                // The formerly latest shard is no
+                // longer a 'recent' shard
+                else
+                {
+                    // Move the formerly latest shard to
+                    // historical storage
+                    if (keepShard(*cur) && separateHistoricalPath)
+                    {
+                        moveShard(*cur);
+                    }
+                }
+
+                cur = boost::none;
+            }
+        }
+    }
+}
+
+boost::optional<bool>
```

Comment:
`optional<bool>` is error prone to use. It's too easy to use the outer `bool` instead of the inner `bool`. I'd recommend replacing the inner `bool` with an enum.

Reply:
Done